packages/compass-query-bar/src/components/query-bar/query-bar.tsx

```diff
@@ -49,74 +61,260 @@ const openQueryHistoryStyles = cx(
+      ),
+    [layout, renderQueryOptionRow]
+  );
+
+  const additionalQueryOptions = useMemo(
+    () =>
+      layout.map((queryOption: string | string[], index: number) =>
+        isQueryOptionsExpanded && index > 0
+          ? renderQueryOptionRow(queryOption, index)
+          : null
+      ),
+    [isQueryOptionsExpanded, layout, renderQueryOptionRow]
+  );
+
+  const onFormSubmit = useCallback(
+    (evt: React.FormEvent) => {
+      evt.preventDefault();
+
+      onApply();
+    },
+    [onApply]
+  );
+
+  return (
+    <form className={queryBarFormStyles} onSubmit={onFormSubmit} noValidate>
+      <div
+        className={cx(
+          queryBarFirstRowStyles,
+          isQueryOptionsExpanded && queryBarFirstRowOpenedStyles
+        )}
+      >
+        {showQueryHistoryButton && (
           <button
             data-test-id="query-history-button"
             onClick={toggleQueryHistory}
             className={openQueryHistoryStyles}
             id="open-query-history"
             aria-label="Open query history"
+            type="button"
           >
             <Icon glyph="Clock" />
             <Icon glyph="CaretDown" />
           </button>
-        </>
-      )}
-      <div className={queryAreaStyles}>Query Area (coming soon)</div>
-      {isQueryOptionsExpanded && <div id="aria-controls">Query Options</div>}
-      <Button
-        data-test-id="query-bar-apply-filter-button"
-        onClick={() => alert('coming soon')}
-        disabled={!isQueryValid}
-        variant="primary"
-        size="small"
-      >
-        {buttonLabel}
-      </Button>
-      <Button
-        aria-label="Reset query"
-        data-test-id="query-bar-reset-filter-button"
-        onClick={() => alert('coming soon')}
-        disabled={queryState !== 'apply'}
-        size="small"
-      >
-        Reset
-      </Button>
-      <MoreOptionsToggle
-        aria-controls="query-options-container"
-        data-testid="query-bar-options-toggle"
-        isExpanded={isQueryOptionsExpanded}
-        onToggleOptions={toggleExpandQueryOptions}
-      />
-    </div>
+        )}
+        {firstRowQueryOptions}
+        <Button
+          aria-label="Reset query"
+          data-test-id="query-bar-reset-filter-button"
+          onClick={onReset}
+          disabled={queryState !== 'apply'}
+          size="small"
+        >
+          Reset
+        </Button>
+        <Button
+          data-test-id="query-bar-apply-filter-button"
+          disabled={!isQueryValid}
+          variant="primary"
+          size="small"
+          type="submit"
+        >
+          {buttonLabel}
+        </Button>
+
+        <MoreOptionsToggle
+          aria-controls="additional-query-options-container"
+          data-testid="query-bar-options-toggle"
+          isExpanded={isQueryOptionsExpanded}
+          onToggleOptions={toggleExpandQueryOptions}
+        />
+      </div>
+      <div id="additional-query-options-container">
+        {additionalQueryOptions}
```

Comment:
If this was a grid and not flex you'd be able to simplify the layout and styles a bunch and this would allow so avoid this annoying misalignment between the row cells

![image](https://user-images.githubusercontent.com/5036933/172556643-7235b2c9-71d4-467e-93ad-d955f48d7793.png)

Reply:
Just pushed an update to make the styles on grid. I couldn't find a way to do grid without introducing some extra complexity, let me know if you can think of some ways to simplify the styles. 
The following constraints were making it a bit tricky:
- Having the learn more link under the less/more options button when possible (larger screen sizes)
- Having the react ace items take up 50% of width
- Having a variable amount of layout items take up the rest of the width while still having their inputs usable (used breakpoints for this one).

I tried out how it looks with large/small screen sizes and the various `layout` configurations cloud uses. Here are the screenshots:
![Screen Shot 2022-06-08 at 7 38 14 PM](https://user-images.githubusercontent.com/1791149/172735688-32e040cf-4833-47c4-9195-fc2db2c701e1.png)
![Screen Shot 2022-06-08 at 7 38 33 PM](https://user-images.githubusercontent.com/1791149/172735691-d70bbda8-47e1-4914-9b00-8982f2d58326.png)
![Screen Shot 2022-06-08 at 7 40 16 PM](https://user-images.githubusercontent.com/1791149/172735692-649839d3-f36f-4ba3-abfe-d29b4b8be7a0.png)
![Screen Shot 2022-06-08 at 7 40 41 PM](https://user-images.githubusercontent.com/1791149/172735693-9135cd1d-f09f-4754-b36f-003fc50856cf.png)
![Screen Shot 2022-06-08 at 7 40 48 PM](https://user-images.githubusercontent.com/1791149/172735695-2400c23c-e62e-4d0b-8eb3-c015e5ec81f7.png)

Reply:
I think you should be able to dynamically build a template area (using `grid-template-areas` css property) based on provided layout to cover most of these cases instead of using template columns, also I think the way you have it implemented right now doesn't cover the requirements you specify (i.e. if I add `skip` as the first item in second row it will take 50% of the space and `project`, `sort` will be 25% because you always have 50% as the first column width) and the first row when has more than a few items is just not rendering correctly due to the space constraints, some items just fall to the next row, not sure if we even want the first row to be a grid here.

Worth nothing that some of the requirements that you mention will just not work with the interface that we allow, nothing stopping me from doing a row like `['filter', 'project', 'sort']` that would make it impossible to have 50% width allocated for ace editor inputs.

I'll once again will say that this `layout` prop interface doesn't make sense to me and if we keep it I guarantee you will have troubles building a robust layout here (there is already cases that don't work correctly). But as you consider changing this out of scope and you have a feeling that grid brings more complexity here feel free to revert it back to flex

Reply:
I spent a few hours today making the query bar options build a dynamic `grid-template-areas` property and I feel it adds too much complexity to cover all of the possible cases (missing one query option for instance). I think we should keep the current interface for now, that way cloud doesn’t have to change their interface, and when we want to do larger changes here we should simplify.

For the basic case we create a `grid-template-areas` like so:
'project project project sort sort sort'
'collation collation skip limit maxTimeMS docs'
When something is missing, or we want to only display one line, everything has to shift or be accounted for.

The current implementation with the overly generic `layout` prop has limitations around what layouts can be passed to it, like you mentioned, passing a row like `[‘filter’, ‘project’, ‘sort’]` will definitely break it and look weird, but that isn’t a current or intended use case. I think that would be apparent in how it breaks when using it so if that’s a use down the line we can accommodate it, and use the information to make a better generalized interface.

This isn’t the best solution, but I also don’t think we should spend the time here to do it correctly to cover use cases we don’t use.

Reply:
Updated to use `grid-template-areas`, built with the query options passed to the QueryBar (replaced the `layout` prop with `queryOptions`). How does the implementation look? Before merging this I'll ping cloud folks on the interface change.